server/src/test/java/org/elasticsearch/common/util/CancellableSingleObjectCacheTests.java

```diff
@@ -222,18 +253,178 @@ protected String getKey(String s) {
+                    }
+
+                    final StepListener<Integer> stepListener = new StepListener<>();
+                    final AtomicBoolean isComplete = new AtomicBoolean();
+                    final AtomicBoolean isCancelled = new AtomicBoolean();
+                    testCache.get(
+                        input,
+                        isCancelled::get,
+                        ActionListener.runBefore(stepListener, () -> assertTrue(isComplete.compareAndSet(false, true)))
+                    );
+
+                    final Runnable next = queue.poll();
+                    if (next != null) {
+                        threadPool.generic().execute(next);
+                    }
+
+                    if (cancel) {
+                        isCancelled.set(true);
+                    }
+
+                    stepListener.whenComplete(len -> {
+                        finishLatch.countDown();
+                        assertThat(len, greaterThanOrEqualTo(input.length()));
+                    }, e -> {
+                        finishLatch.countDown();
+                        if (e instanceof TaskCancelledException) {
+                            assertTrue(cancel);
+                        } else {
+                            throw new AssertionError("unexpected", e);
+                        }
+                    });
+                });
+            }
+
+            for (int i = 0; i < 10; i++) {
+                threadPool.generic().execute(Objects.requireNonNull(queue.poll()));
+            }
+
+            startLatch.countDown();
+            assertTrue(finishLatch.await(10, TimeUnit.SECONDS));
+        } finally {
+            ThreadPool.terminate(threadPool, 10, TimeUnit.SECONDS);
+        }
+    }
+
+    public void testForegroundRefreshCanBeCancelled() throws InterruptedException {
+
+        final Runnable awaitBarrier = new Runnable() {
+            final CyclicBarrier barrier = new CyclicBarrier(2);
+
+            @Override
+            public void run() {
+                try {
+                    barrier.await(10, TimeUnit.SECONDS);
+                } catch (InterruptedException | BrokenBarrierException | TimeoutException e) {
+                    throw new AssertionError("unexpected", e);
+                }
+            }
+        };
+
+        final AtomicInteger successfulChecksRemaining = new AtomicInteger(between(2, 10));
+
+        final CancellableSingleObjectCache<String, String, Integer> testCache = new CancellableSingleObjectCache<>() {
+            @Override
+            protected void refresh(String s, Runnable ensureNotCancelled, ActionListener<Integer> listener, BooleanSupplier isFreshCheck) {
+                ActionListener.completeWith(listener, () -> {
+                    while (successfulChecksRemaining.get() > 0) {
+                        ensureNotCancelled.run();
+                    }
+                    awaitBarrier.run();
+                    awaitBarrier.run();
```

Comment:
The barrier usage here is quite involved, I wonder if we should add a comment. Took me a while to get to the fact that the second `awaitBarrier` in the cancelled refresh is matched by the first `awaitBarrier` in the successful one. I also wonder if that complexity is necessary?

Reply:
Coming at this with a fresh brain I see some simplifications now, see 16da0ae7431e493d7569f5ab75aa877dfe1e5f8e. Also added comments to show the order in which the barriers are hit on the respective threads.

Reply:
Thanks, much clearer with the comments.